File: sdk/src/types/manifest.ts
/**
 * Plugin manifest types (plugin.hone.json).
 */

import type { ConfigSchema } from './config';

export interface PluginCapabilities {
  'editor.read'?: boolean;
  'editor.write'?: boolean;
  'editor.decorations'?: boolean;
  'filesystem.read'?: string[];
  'filesystem.write'?: boolean;
  'network'?: boolean;
  'process.spawn'?: string[];
  'terminal'?: boolean;
  'ui.panel'?: boolean;
  'ui.statusbar'?: boolean;
  'ui.gutter'?: boolean;
  'ui.commandPalette'?: boolean;
  'ui.contextMenu'?: boolean;
  'ui.notifications'?: boolean;
  'ui.webview'?: boolean;
}

export type HookName =
  | 'onDocumentFormat'
  | 'onDocumentSave'
  | 'onDocumentOpen'
  | 'onSelectionChange'
  | 'onCodeAction'
  | 'onHover'
  | string; // onCommand:* patterns

export interface PluginManifest {
  name: string;
  displayName: string;
  version: string;
  author: string;
  license: string;
  description: string;
  entry: string;
  capabilities: PluginCapabilities;
  hooks: HookName[];
  configSchema?: ConfigSchema;
  hone: string;
  repository?: string;
  icon?: string;
  perryVersion?: string;
}

/** Plugin execution tier derived from capabilities. */
export type PluginTier = 1 | 2 | 3;

/**
 * Derive the plugin tier from its declared capabilities.
 *
 * Tier 1 (InProcess): UI-only, declarative data (themes, keymaps).
 * Tier 2 (PluginHost): editor/fs-read/ui, shared process.
 * Tier 3 (IsolatedProcess): network/fs-write/process.spawn/terminal/webview.
 */
export function deriveTier(caps: PluginCapabilities): PluginTier {
  // Tier 3: any "dangerous" capability
  if (caps.network === true) return 3;
  if (caps['filesystem.write'] === true) return 3;
  if (caps.terminal === true) return 3;
  if (caps['ui.webview'] === true) return 3;
  const spawn = caps['process.spawn'];
  if (spawn !== undefined && spawn.length > 0) return 3;

  // Tier 2: any editor/fs-read/ui capability
  if (caps['editor.read'] === true) return 2;
  if (caps['editor.write'] === true) return 2;
  if (caps['editor.decorations'] === true) return 2;
  const fsRead = caps['filesystem.read'];
  if (fsRead !== undefined && fsRead.length > 0) return 2;
  if (caps['ui.panel'] === true) return 2;
  if (caps['ui.statusbar'] === true) return 2;
  if (caps['ui.gutter'] === true) return 2;
  if (caps['ui.commandPalette'] === true) return 2;
  if (caps['ui.contextMenu'] === true) return 2;
  if (caps['ui.notifications'] === true) return 2;

  // Tier 1: UI-only (themes, keymaps, declarative data)
  return 1;
}

/** Validation error for a plugin manifest. */
export interface ManifestError {
  field: string;
  message: string;
}

/**
 * Validate a plugin manifest object.
 * Returns an array of errors (empty if valid).
 */
export function validateManifest(data: Record<string, unknown>): ManifestError[] {
  const errors: ManifestError[] = [];

  if (typeof data.name !== 'string' || data.name.length === 0) {
    errors.push({ field: 'name', message: 'name is required and must be a non-empty string' });
  }
  if (typeof data.displayName !== 'string' || data.displayName.length === 0) {
    errors.push({ field: 'displayName', message: 'displayName is required and must be a non-empty string' });
  }
  if (typeof data.version !== 'string' || data.version.length === 0) {
    errors.push({ field: 'version', message: 'version is required and must be a non-empty string' });
  }
  if (typeof data.author !== 'string' || data.author.length === 0) {
    errors.push({ field: 'author', message: 'author is required and must be a non-empty string' });
  }
  if (typeof data.license !== 'string' || data.license.length === 0) {
    errors.push({ field: 'license', message: 'license is required and must be a non-empty string' });
  }
  if (typeof data.description !== 'string') {
    errors.push({ field: 'description', message: 'description is required and must be a string' });
  }
  if (typeof data.entry !== 'string' || data.entry.length === 0) {
    errors.push({ field: 'entry', message: 'entry is required and must be a non-empty string' });
  }

  // Capabilities validation
  if (data.capabilities === undefined || data.capabilities === null || typeof data.capabilities !== 'object') {
    errors.push({ field: 'capabilities', message: 'capabilities is required and must be an object' });
  } else {
    const caps = data.capabilities as Record<string, unknown>;
    const fsRead = caps['filesystem.read'];
    if (fsRead !== undefined && !Array.isArray(fsRead)) {
      errors.push({ field: 'capabilities.filesystem.read', message: 'filesystem.read must be an array of glob strings' });
    }
    const spawn = caps['process.spawn'];
    if (spawn !== undefined && !Array.isArray(spawn)) {
      errors.push({ field: 'capabilities.process.spawn', message: 'process.spawn must be an array of binary names' });
    }
  }

  // Hooks validation
  if (!Array.isArray(data.hooks)) {
    errors.push({ field: 'hooks', message: 'hooks is required and must be an array' });
  }

  // Engine version
  if (typeof data.hone !== 'string' || data.hone.length === 0) {
    errors.push({ field: 'hone', message: 'hone engine version is required' });
  }

  return errors;
}

/**
 * Parse a plugin manifest from a JSON string.
 */
export function parsePluginManifest(json: string): { manifest: PluginManifest | null; errors: ManifestError[] } {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(json);
  } catch {
    return { manifest: null, errors: [{ field: 'root', message: 'Invalid JSON' }] };
  }

  const errors = validateManifest(data);
  if (errors.length > 0) {
    return { manifest: null, errors: errors };
  }

  const caps = data.capabilities as Record<string, unknown>;
  const manifest: PluginManifest = {
    name: data.name as string,
    displayName: data.displayName as string,
    version: data.version as string,
    author: data.author as string,
    license: data.license as string,
    description: data.description as string,
    entry: data.entry as string,
    capabilities: {
      'editor.read': caps['editor.read'] === true,
      'editor.write': caps['editor.write'] === true,
      'editor.decorations': caps['editor.decorations'] === true,
      'filesystem.read': Array.isArray(caps['filesystem.read']) ? caps['filesystem.read'] as string[] : undefined,
      'filesystem.write': caps['filesystem.write'] === true,
      'network': caps['network'] === true,
      'process.spawn': Array.isArray(caps['process.spawn']) ? caps['process.spawn'] as string[] : undefined,
      'terminal': caps['terminal'] === true,
      'ui.panel': caps['ui.panel'] === true,
      'ui.statusbar': caps['ui.statusbar'] === true,
      'ui.gutter': caps['ui.gutter'] === true,
      'ui.commandPalette': caps['ui.commandPalette'] === true,
      'ui.contextMenu': caps['ui.contextMenu'] === true,
      'ui.notifications': caps['ui.notifications'] === true,
      'ui.webview': caps['ui.webview'] === true,
    },
    hooks: data.hooks as string[],
    configSchema: data.configSchema as ConfigSchema | undefined,
    hone: data.hone as string,
    repository: data.repository as string | undefined,
    icon: data.icon as string | undefined,
    perryVersion: data.perryVersion as string | undefined,
  };

  return { manifest: manifest, errors: [] };
}

File: host/src/manifest.rs
//! Plugin manifest parsing — plugin.hone.json schema.

use serde::Deserialize;
use std::collections::HashMap;
use std::path::Path;

/// Parsed plugin manifest from plugin.hone.json.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PluginManifest {
    pub name: String,
    pub display_name: String,
    pub version: String,
    pub author: String,
    pub license: String,
    pub description: String,
    pub entry: String,
    #[serde(default)]
    pub capabilities: Capabilities,
    #[serde(default)]
    pub hooks: Vec<String>,
    #[serde(default)]
    pub config_schema: HashMap<String, ConfigEntry>,
    pub hone: String,
    pub repository: Option<String>,
    pub icon: Option<String>,
    pub perry_version: Option<String>,
}

/// Declared capabilities determining which Host API functions are available.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Capabilities {
    #[serde(default, rename = "editor.read")]
    pub editor_read: bool,
    #[serde(default, rename = "editor.write")]
    pub editor_write: bool,
    #[serde(default, rename = "editor.decorations")]
    pub editor_decorations: bool,
    #[serde(default, rename = "filesystem.read")]
    pub filesystem_read: Option<Vec<String>>,
    #[serde(default, rename = "filesystem.write")]
    pub filesystem_write: bool,
    #[serde(default)]
    pub network: bool,
    #[serde(default, rename = "process.spawn")]
    pub process_spawn: Option<Vec<String>>,
    #[serde(default)]
    pub terminal: bool,
    #[serde(default, rename = "ui.panel")]
    pub ui_panel: bool,
    #[serde(default, rename = "ui.statusbar")]
    pub ui_statusbar: bool,
    #[serde(default, rename = "ui.gutter")]
    pub ui_gutter: bool,
    #[serde(default, rename = "ui.commandPalette")]
    pub ui_command_palette: bool,
    #[serde(default, rename = "ui.contextMenu")]
    pub ui_context_menu: bool,
    #[serde(default, rename = "ui.notifications")]
    pub ui_notifications: bool,
    #[serde(default, rename = "ui.webview")]
    pub ui_webview: bool,
}

/// A configuration entry in the plugin's config schema.
#[derive(Debug, Clone, Deserialize)]
pub struct ConfigEntry {
    #[serde(rename = "type")]
    pub value_type: String,
    pub default: Option<serde_json::Value>,
    pub description: Option<String>,
    #[serde(rename = "enum")]
    pub enum_values: Option<Vec<String>>,
}

/// Manifest validation error.
#[derive(Debug)]
pub struct ManifestError {
    pub field: String,
    pub message: String,
}

impl std::fmt::Display for ManifestError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.field, self.message)
    }
}

/// Parse a plugin manifest from a JSON string.
pub fn parse_manifest(json: &str) -> Result<PluginManifest, Vec<ManifestError>> {
    let manifest: PluginManifest = serde_json::from_str(json).map_err(|e| {
        vec![ManifestError {
            field: "root".to_string(),
            message: format!("Invalid JSON: {}", e),
        }]
    })?;

    let errors = validate_manifest(&manifest);
    if !errors.is_empty() {
        return Err(errors);
    }

    Ok(manifest)
}

/// Parse a plugin manifest from a file path.
pub fn parse_manifest_file(path: &Path) -> Result<PluginManifest, Vec<ManifestError>> {
    let json = std::fs::read_to_string(path).map_err(|e| {
        vec![ManifestError {
            field: "file".to_string(),
            message: format!("Cannot read file: {}", e),
        }]
    })?;
    parse_manifest(&json)
}

/// Validate a parsed manifest.
fn validate_manifest(m: &PluginManifest) -> Vec<ManifestError> {
    let mut errors = Vec::new();

    if m.name.is_empty() {
        errors.push(ManifestError {
            field: "name".to_string(),
            message: "name must be non-empty".to_string(),
        });
    }
    if m.display_name.is_empty() {
        errors.push(ManifestError {
            field: "displayName".to_string(),
            message: "displayName must be non-empty".to_string(),
        });
    }
    if m.version.is_empty() {
        errors.push(ManifestError {
            field: "version".to_string(),
            message: "version must be non-empty".to_string(),
        });
    }
    if m.author.is_empty() {
        errors.push(ManifestError {
            field: "author".to_string(),
            message: "author must be non-empty".to_string(),
        });
    }
    if m.license.is_empty() {
        errors.push(ManifestError {
            field: "license".to_string(),
            message: "license must be non-empty".to_string(),
        });
    }
    if m.hone.is_empty() {
        errors.push(ManifestError {
            field: "hone".to_string(),
            message: "hone engine version must be non-empty".to_string(),
        });
    }

    errors
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_valid_manifest() {
        let json = r#"{
            "name": "test-plugin",
            "displayName": "Test Plugin",
            "version": "1.0.0",
            "author": "Test Author",
            "license": "MIT",
            "description": "A test plugin",
            "entry": "TestPlugin",
            "capabilities": { "editor.read": true, "ui.statusbar": true },
            "hooks": ["onDocumentOpen"],
            "hone": ">=0.1.0"
        }"#;
        let manifest = parse_manifest(json).unwrap();
        assert_eq!(manifest.name, "test-plugin");
        assert!(manifest.capabilities.editor_read);
        assert!(manifest.capabilities.ui_statusbar);
        assert!(!manifest.capabilities.network);
        assert_eq!(manifest.hooks.len(), 1);
    }

    #[test]
    fn parse_minimal_manifest() {
        let json = r#"{
            "name": "min",
            "displayName": "Min",
            "version": "0.1.0",
            "author": "A",
            "license": "MIT",
            "description": "",
            "entry": "",
            "capabilities": {},
            "hooks": [],
            "hone": ">=0.1.0"
        }"#;
        let manifest = parse_manifest(json).unwrap();
        assert_eq!(manifest.name, "min");
    }

    #[test]
    fn parse_invalid_json() {
        let result = parse_manifest("not json");
        assert!(result.is_err());
    }

    #[test]
    fn parse_missing_name() {
        let json = r#"{
            "displayName": "X",
            "version": "1.0.0",
            "author": "A",
            "license": "MIT",
            "description": "",
            "entry": "X",
            "capabilities": {},
            "hooks": [],
            "hone": ">=0.1.0"
        }"#;
        // serde will fail because name is required
        let result = parse_manifest(json);
        assert!(result.is_err());
    }

    #[test]
    fn parse_filesystem_read_globs() {
        let json = r#"{
            "name": "fs-plugin",
            "displayName": "FS Plugin",
            "version": "1.0.0",
            "author": "A",
            "license": "MIT",
            "description": "desc",
            "entry": "FsPlugin",
            "capabilities": { "filesystem.read": ["**/*.ts", "**/*.js"] },
            "hooks": [],
            "hone": ">=0.1.0"
        }"#;
        let manifest = parse_manifest(json).unwrap();
        let globs = manifest.capabilities.filesystem_read.as_ref().unwrap();
        assert_eq!(globs.len(), 2);
        assert_eq!(globs[0], "**/*.ts");
    }

    #[test]
    fn parse_process_spawn_allowlist() {
        let json = r#"{
            "name": "lint",
            "displayName": "Lint",
            "version": "1.0.0",
            "author": "A",
            "license": "MIT",
            "description": "desc",
            "entry": "LintPlugin",
            "capabilities": { "process.spawn": ["eslint", "prettier"] },
            "hooks": [],
            "hone": ">=0.1.0"
        }"#;
        let manifest = parse_manifest(json).unwrap();
        let bins = manifest.capabilities.process_spawn.as_ref().unwrap();
        assert_eq!(bins.len(), 2);
    }

    #[test]
    fn parse_config_schema() {
        let json = r#"{
            "name": "cfg",
            "displayName": "Cfg",
            "version": "1.0.0",
            "author": "A",
            "license": "MIT",
            "description": "",
            "entry": "CfgPlugin",
            "capabilities": {},
            "hooks": [],
            "hone": ">=0.1.0",
            "configSchema": {
                "printWidth": { "type": "number", "default": 80, "description": "Line width" },
                "semi": { "type": "boolean", "default": true }
            }
        }"#;
        let manifest = parse_manifest(json).unwrap();
        assert_eq!(manifest.config_schema.len(), 2);
        assert_eq!(manifest.config_schema["printWidth"].value_type, "number");
    }
}
